src/high-level/include/iDynTree/KinDynComputations.h

```diff
@@ -328,20 +408,60 @@ class KinDynComputations {
                        iDynTree::VectorDynSize &s_dot,
                        iDynTree::Vector3& world_gravity);
 
+    bool getRobotState(const iDynTree::Span<double>& world_p_base,
+                       const iDynTree::MatrixView<double>& world_R_base,
+                       const iDynTree::Span<double>& s,
+                       const iDynTree::Span<double>& base_velocity,
+                       const iDynTree::Span<double>& s_dot,
+                       const iDynTree::Span<double>& world_gravity);
+
+    void getRobotState(const iDynTree::Span<double>& s,
+                       const iDynTree::Span<double>& s_dot,
+                       const iDynTree::Span<double>& world_gravity);
+
+
     /**
      * Access the robot state.
      */
-    iDynTree::Transform getWorldBaseTransform();
-    iDynTree::Twist     getBaseTwist();
+    iDynTree::Transform getWorldBaseTransform() const;
+    bool getWorldBaseTransform(const iDynTree::Span<double>& world_p_base,
+                               const iDynTree::MatrixView<double>& world_R_base) const;
+
+    iDynTree::Twist getBaseTwist() const;
+    bool getBaseTwist(const iDynTree::Span<double> & base_velocity) const;
+
+    bool getJointPos(iDynTree::VectorDynSize &q) const;
+
+    /**
+     * Get the n joint position of the model.
+     * @warning the Span object should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
+     * @return true if all went well, false otherwise.
+     */
+    bool getJointPos(const iDynTree::Span<double> &q) const;
 
-    bool getJointPos(iDynTree::VectorDynSize &q);
-    bool getJointVel(iDynTree::VectorDynSize &dq);
+    bool getJointVel(iDynTree::VectorDynSize &dq) const;
```

Comment:
If:
- performance tests do not show regression
- you do not have a particular reason to use this,

I suggest to deprecate the explicit vector version and keep only methods accepting `Span` or `MatrixView`: cleaner interface (of course, where it is possible, i.e. not in `getWorldTransform`)

Reply:
The explicit vector version has always had a slightly different semantics: it resize the vector to be of the correct size.

Reply:
I am fine in keeping things as they are in this PR, but I strongly suggest revisiting this later (but not too late). 
API bloating is a problem. 
Deprecating and suggesting the user to pass a pre-allocated span should not be a problem. 
Open an issue to track this.

Reply:
>  Deprecating and suggesting the user to pass a pre-allocated span should not be a problem.

This API is quite new, and in general we do not know if the use of Span/MatrixView is going to create unforeseeable problems. For example, I don't know if the implicit conversion from VectorDynSize to Span works correctly in the Matlab/Python bindings as it works in C++. So I opened an issue to track this in https://github.com/robotology/idyntree/issues/737, but I personally would like to try this API extensively in the wild before deprecating and removing the old methods.

Reply:
@traversaro I read this last comment after I wrote https://github.com/robotology/idyntree/issues/737#issuecomment-691930285, I think that what I wrotefor Python also applied to Matlab, good point. We have a lot of code that use the Matlab bindings that I'm not sure we want to break. Python situation is instead different.